app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js

```diff
@@ -0,0 +1,182 @@
+    )
+  }
+
+  const { origin } = req
+
+  const {
+    chainId,
+    chainName = null,
+    blockExplorerUrls = null,
+    nativeCurrency = null,
+    rpcUrls,
+  } = req.params[0]
+
+  const otherKeys = Object.keys(
+    omit(req.params[0], [
+      'chainId',
+      'chainName',
+      'blockExplorerUrls',
+      'iconUrl',
+      'rpcUrls',
+      'nativeCurrency',
+    ]),
+  )
+
+  if (otherKeys.length > 0) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Received unexpected keys on object parameter. Unsupported keys:\n${otherKeys}`,
+      }),
+    )
+  }
+
+  const firstValidRPCUrl = Array.isArray(rpcUrls)
+    ? rpcUrls.find((rpcUrl) => validUrl.isHttpsUri(rpcUrl))
+    : null
+
+  const firstValidBlockExplorerUrl =
+    blockExplorerUrls !== null && Array.isArray(blockExplorerUrls)
+      ? blockExplorerUrls.find((blockExplorerUrl) =>
+          validUrl.isHttpsUri(blockExplorerUrl),
+        )
+      : null
+
+  if (!firstValidRPCUrl) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Expected an array with at least one valid string HTTPS url 'rpcUrls', Received:\n${rpcUrls}`,
+      }),
+    )
+  }
+
+  const _chainId = typeof chainId === 'string' && chainId.toLowerCase()
+  if (!isPrefixedFormattedHexString(_chainId)) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Expected 0x-prefixed, unpadded, non-zero hexadecimal string 'chainId'. Received:\n${chainId}`,
+      }),
+    )
+  }
+
+  // TODO: Check specified chain ID against endpoint chain ID, as in custom network form.
+
+  // TODO: Disallow adding default networks
+  // if (DEFAULT_NETWORK_MAP[_chainId]) {
+  //   return end(ethErrors.rpc.invalidParams({
+  //     message: `May not specify default MetaMask chain.`,
+  //   }))
+  // }
+
+  if (typeof chainName !== 'string' || !chainName) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Expected non-empty string 'chainName'. Received:\n${chainName}`,
+      }),
+    )
+  }
+  const _chainName =
+    chainName.length > 100 ? chainName.substring(0, 100) : chainName
+
+  if (nativeCurrency !== null && typeof nativeCurrency !== 'object') {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Expected null or object 'nativeCurrency'. Received:\n${nativeCurrency}`,
+      }),
+    )
+  }
+
+  if (
+    nativeCurrency !== null &&
+    (!nativeCurrency.decimals || nativeCurrency.decimals !== 18)
+  ) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Expected the number 18 for 'nativeCurrency.decimals' when 'nativeCurrency' is provided. Received: ${nativeCurrency.decimals}`,
+      }),
+    )
+  }
+
+  if (
+    nativeCurrency !== null &&
+    (!nativeCurrency.symbol || typeof nativeCurrency.symbol !== 'string')
+  ) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Expected a string 'nativeCurrency.symbol'. Received: ${nativeCurrency.symbol}`,
+      }),
+    )
+  }
```

Comment:
- We can move the `.decimals` and `.symbol` validation into a single `nativeCurrency !== null` condition.
- We can just do `nativeCurrency.decimals !== 18`

Reply:
done